src/plugins/visualizations/public/utils/saved_visualize_utils.ts

```diff
@@ -0,0 +1,400 @@
+    } catch (error: any) {
+      throw error;
+    }
+  }
+
+  if (resp.references && resp.references.length > 0) {
+    injectReferences(savedObject, resp.references);
+  }
+
+  if (services.savedObjectsTagging) {
+    savedObject.tags = services.savedObjectsTagging.ui.getTagIdsFromReferences(resp.references);
+  }
+
+  savedObject.visState = await updateOldState(savedObject.visState);
+  if (savedObject.searchSourceFields?.index) {
+    await services.dataViews.get(savedObject.searchSourceFields.index as any);
+  }
+
+  return savedObject;
+}
+
+export async function saveVisualization(
+  savedObject: VisSavedObject,
+  { confirmOverwrite = false, isTitleDuplicateConfirmed = false, onTitleDuplicate }: SaveVisOptions,
+  services: {
+    savedObjectsClient: SavedObjectsClientContract;
+    overlays: OverlayStart;
+    savedObjectsTagging?: SavedObjectsTaggingApi;
+  }
+) {
+  // Save the original id in case the save fails.
+  const originalId = savedObject.id;
+  // Read https://github.com/elastic/kibana/issues/9056 and
+  // https://github.com/elastic/kibana/issues/9012 for some background into why this copyOnSave variable
+  // exists.
+  // The goal is to move towards a better rename flow, but since our users have been conditioned
+  // to expect a 'save as' flow during a rename, we are keeping the logic the same until a better
+  // UI/UX can be worked out.
+  if (savedObject.copyOnSave) {
+    delete savedObject.id;
+  }
+
+  const attributes: SavedObjectAttributes = {
+    visState: JSON.stringify(savedObject.visState),
+    title: savedObject.title,
+    uiStateJSON: savedObject.uiStateJSON,
+    description: savedObject.description,
+    savedSearchId: savedObject.savedSearchId,
+    version: savedObject.version,
+  };
+  let references: SavedObjectReference[] = [];
+
+  if (savedObject.searchSource) {
+    const { searchSourceJSON, references: searchSourceReferences } =
+      savedObject.searchSource.serialize();
+    attributes.kibanaSavedObjectMeta = { searchSourceJSON };
+    references.push(...searchSourceReferences);
+  }
+
+  if (savedObject.searchSourceFields) {
+    const [searchSourceFields, searchSourceReferences] = extractSearchSourceReferences(
+      savedObject.searchSourceFields
+    );
+    const searchSourceJSON = JSON.stringify(searchSourceFields);
+    attributes.kibanaSavedObjectMeta = { searchSourceJSON };
+    references.push(...searchSourceReferences);
+  }
+
+  if (services.savedObjectsTagging) {
+    references = services.savedObjectsTagging.ui.updateTagsReferences(
+      references,
+      savedObject.tags || []
+    );
+  }
+
+  const extractedRefs = extractReferences({ attributes, references });
+
+  if (!extractedRefs.references) {
+    throw new Error('References not returned from extractReferences');
+  }
+
+  try {
+    await checkForDuplicateTitle(
+      savedObject as any,
+      isTitleDuplicateConfirmed,
+      onTitleDuplicate,
+      services as any
+    );
+    savedObject.isSaving = true;
```

Comment:
please check it, looks like it's do nothing

Reply:
agree, I will remove it